ouroboros-consensus/src/ouroboros-consensus/Ouroboros/Consensus/HardFork/Combinator/Ledger/Query.hs

```diff
@@ -211,43 +315,90 @@ getHardForkQuery q k1 k2 k3 = case q of
   Current era queries
 -------------------------------------------------------------------------------}
 
-data QueryIfCurrent :: [Type] -> Type -> Type where
-  QZ :: BlockQuery x result      -> QueryIfCurrent (x ': xs) result
-  QS :: QueryIfCurrent xs result -> QueryIfCurrent (x ': xs) result
+type QueryIfCurrent :: [Type] -> QueryFootprint -> Type -> Type
+data QueryIfCurrent xs footprint result where
+  QZ :: BlockQuery     x  footprint result -> QueryIfCurrent (x ': xs) footprint result
+  QS :: QueryIfCurrent xs footprint result -> QueryIfCurrent (x ': xs) footprint result
 
-deriving instance All SingleEraBlock xs => Show (QueryIfCurrent xs result)
+deriving instance All SingleEraBlock xs => Show (QueryIfCurrent xs footprint result)
 
-instance All SingleEraBlock xs => ShowQuery (QueryIfCurrent xs) where
+instance All SingleEraBlock xs => ShowQuery (QueryIfCurrent xs footprint) where
   showResult (QZ qry) = showResult qry
   showResult (QS qry) = showResult qry
 
-instance All SingleEraBlock xs => SameDepIndex (QueryIfCurrent xs) where
-  sameDepIndex (QZ qry) (QZ qry') = sameDepIndex qry qry'
-  sameDepIndex (QS qry) (QS qry') = sameDepIndex qry qry'
-  sameDepIndex _        _         = Nothing
+instance All SingleEraBlock xs => SameDepIndex2 (QueryIfCurrent xs) where
+  sameDepIndex2 (QZ qry) (QZ qry') = sameDepIndex2 qry qry'
+  sameDepIndex2 (QS qry) (QS qry') = sameDepIndex2 qry qry'
+  sameDepIndex2 _        _         = Nothing
 
 interpretQueryIfCurrent ::
      forall result xs. All SingleEraBlock xs
   => NP ExtLedgerCfg xs
-  -> QueryIfCurrent xs result
-  -> NS ExtLedgerState xs
+  -> QueryIfCurrent xs QFNoTables result
+  -> NS (Flip ExtLedgerState EmptyMK) xs
   -> HardForkQueryResult xs result
 interpretQueryIfCurrent = go
   where
     go :: All SingleEraBlock xs'
        => NP ExtLedgerCfg xs'
-       -> QueryIfCurrent xs' result
-       -> NS ExtLedgerState xs'
+       -> QueryIfCurrent xs' QFNoTables result
+       -> NS (Flip ExtLedgerState EmptyMK) xs'
        -> HardForkQueryResult xs' result
-    go (c :* _)  (QZ qry) (Z st) =
-        Right $ answerBlockQuery c qry st
+    go (c :* _)  (QZ qry) (Z (Flip st)) =
+        Right $ answerPureBlockQuery c qry st
     go (_ :* cs) (QS qry) (S st) =
         first shiftMismatch $ go cs qry st
     go _         (QZ qry) (S st) =
-        Left $ MismatchEraInfo $ ML (queryInfo qry) (hcmap proxySingle ledgerInfo st)
-    go _         (QS qry) (Z st) =
+        Left $ MismatchEraInfo $ ML (queryInfo qry) (hcmap proxySingle (ledgerInfo . unFlip) st)
+    go _         (QS qry) (Z (Flip st)) =
         Left $ MismatchEraInfo $ MR (hardForkQueryInfo qry) (ledgerInfo st)
 
+interpretQueryIfCurrentOne ::
```

Comment:
No you cannot build a forker from a pure ledger state. A forker needs the tables, or at least the ability to read those from the storage.

Reply:
> No you cannot build a forker from a pure ledger state. A forker needs the tables, or at least the ability to read those from the storage.

Looks to me like you could satisfy the `ReadOnlyForker` interface, but that it would be a degenerate `Forker`: it won't have any tables because it's constructed from `EmptyMK`.

It's probably not worth it though to have a `ReadOnlyForker` double as a handle to a ledger state with the full UTXO *and* a ledger state with a subset of the UTXO. It could introduce run-time bugs if we're not careful, unless `Forker` also gets a type parameter like a mapkind. 

TDR: the current approach has some duplication, but it has more static guarantees